library/std/src/net/parser.rs

```diff
@@ -84,29 +107,20 @@ impl<'a> Parser<'a> {
         })
     }
 
-    // Read a single digit in the given radix. For instance, 0-9 in radix 10;
-    // 0-9A-F in radix 16.
-    fn read_digit(&mut self, radix: u32) -> Option<u32> {
-        self.read_atomically(move |p| p.read_char()?.to_digit(radix))
-    }
-
     // Read a number off the front of the input in the given radix, stopping
-    // at the first non-digit character or eof. Fails if the number has more
-    // digits than max_digits, or the value is >= upto, or if there is no number.
-    fn read_number(&mut self, radix: u32, max_digits: u32, upto: u32) -> Option<u32> {
+    // at the first non-digit character or eof.
+    fn read_number<T: ReadNumberHelper>(&mut self, radix: u32) -> Option<T> {
         self.read_atomically(move |p| {
-            let mut result = 0;
-            let mut digit_count = 0;
-
-            while let Some(digit) = p.read_digit(radix) {
-                result = (result * radix) + digit;
-                digit_count += 1;
-                if digit_count > max_digits || result >= upto {
-                    return None;
-                }
+            let mut result = T::ZERO;
+            let mut parsed = false;
+
+            while let Some(digit) = p.read_atomically(|p| p.read_char()?.to_digit(radix)) {
```

Comment:
Do we need a second `read_atomically`?

Reply:
Yes, because if the next thing we read is not a digit, we want to rewind the parser.